### ISSUE TITLE
PyTorch memory leak reference cycle in for loop, Mac M1 

### ISSUE THREAD

**Issue by Davidvandijcke** [pytorch/pytorch#91368]
### 🐛 Describe the bug

I am facing a memory leak when iteratively updating tensors in PyTorch on my Mac M1 GPU using the PyTorch mps interface. The following is a minimal reproducible example that replicates the behavior:

```python
import torch 

def leak_example(p1, device):
    
    t1 = torch.rand_like(p1, device = device) # torch.cat((torch.diff(ubar.detach(), dim=0).detach().clone(), torch.zeros_like(ubar.detach()[:1,:,:,:], dtype = torch.float32)), dim = 0)
    u1 = p1.detach() + 2 * (t1.detach())
    
    B = torch.rand_like(u1, device = device)
    mask = u1 < B
    
    a1 = u1.detach().clone()
    a1[~mask] = torch.rand_like(a1)[~mask]
    return a1

if torch.cuda.is_available(): # cuda gpus
    device = torch.device("cuda")
elif torch.backends.mps.is_available(): # mac gpus
    device = torch.device("mps")
torch.set_grad_enabled(False)
        
p1 = torch.rand(5, 5, 224, 224, device = device)
for i in range(10000):
    p1 = leak_example(p1, device)    
```

My Mac's GPU memory steadily grows when I execute this loop. I have tried running it on a CUDA GPU in Google Colab and the output of `print(torch.cuda.memory_allocated()/1024**2)` remains constant, though the GPU's Active memory, Non-releasable memory, and Allocated memory do appear to increase as the loop progresses (the output of `torch.cuda.memory_summary()`).

I have tried detaching and cloning the tensors and using weakrefs, to no avail. Interestingly, if I don't reassign the output of `leak_example` to `p1`, the behavior disappears, so it really seems related to the recursive assignment. Also, if I drop the `torch.diff()` assignment, the memory increase stalls out at some point, so it may be related to this function. Does anyone have any idea how I could resolve this?

### Versions

PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

**Comment by Davidvandijcke** [comment]
On further inspection, it actually seems like the memory leak is mostly driven by the masked assignment, `a1[~mask] = torch.rand_like(a1)[~mask]. Commenting that out makes the behavior disappear.

**Comment by razarmehr** [comment]
Hi @Davidvandijcke. I tested with the latest 'nightly' build and memory seems stable with the test you provided. There's a chance that the recently-merged #91348 could've fixed this. Can you please try the latest nightly build? Thanks.

**Comment by sbond75** [comment]
I tried on my M1 Max and got a leak on the latest nightly build (version 2.0.0.dev20230105) just now.

**Comment by sbond75** [comment]
I also tried the commit [688e351970776120295843e88e407e3cd6e95bbc](https://github.com/pytorch/pytorch/commit/688e351970776120295843e88e407e3cd6e95bbc) and found that the leak still happens, and I put it through Xcode Instruments to try to see where the leak is happening. It appears to be from calls to IOGPUMetalBuffer initWithDevice shown in this screenshot.
![Screen Shot 2023-01-06 at 5 45 40 PM](https://user-images.githubusercontent.com/43617712/211112634-007f8f14-01ec-4f93-9011-8497086d4cf6.png)

A Ctrl-F for "initWithDevice" on the commit log for the commit I used shows a few calls to initWithDevice, not sure if they are related.

**Comment by jyc** [comment]
I can confirm that the test script leaks memory on my M1 Max.
